### PR TITLE
Gimbaled PrecLand

### PR DESCRIPTION
**Pull request by squilter**
Adds support for `frame` within the `LANDING_TARGET` message.  This allows a companion computer to use the video feed from a gimbaled camera for landing.

@djnugent Would you mind testing your cool simulation to make sure I didn't break your thing?  Pretty sure I didn't, but I just want to be sure.

### CODE REVIEW

**Comment by djnugent** [comment]
Looks good to me. My simulator uses a hard mount camera. I'll have to modify it be a gimballed camera. I'll try to test it in the next few days.

**Comment by djnugent** [comment]
Is there any push to have a virtual world associated with SITL? It would be extremely handy for testing vision work. Tighter integration with Gazebo seems like a good option.

**Comment by squilter** [comment]
I just meant test the simulator with your hard mount.  No need to make hardware mods.  I want to make sure that this doesn't break you hard-mount stuff.  It's probably a lot of work to set up a gimbaled precland, so I'm not asking you to do that.

Not sure about #2, but I agree that it would be really awesome if there were some better way of doing CV stuff with SITL.

**Comment by djnugent** [comment]
I see. Yeah I'll get to it later this week. SITL is currently broken on my machine. Need to reconfigure my environment.

**Comment by squilter** [comment]
thanks

**Comment by gmorph** [comment]
I've added the "in progress" label so we know it being worked on by you Daniel.  Once you've tested it let us know how you went.
Thanks, Grant.

File: libraries/AC_PrecLand/AC_PrecLand.cpp [hunk 1]
```diff
@@ -136,18 +136,23 @@ void AC_PrecLand::calc_angles_and_pos(float alt_above_terrain_cm)
...
+    
+    if(_backend->get_frame_of_reference() == MAV_FRAME_LOCAL_NED){
+        _ef_angle_to_target.x = _angle_to_target.x;
+        _ef_angle_to_target.y = _angle_to_target.y;
+    }else{ // assume MAV_FRAME_BODY_NED (i.e. a hard-mounted sensor)
+        // subtract vehicle lean angles
+        float x_rad = _angle_to_target.x - _ahrs.roll;
+        float y_rad = -_angle_to_target.y + _ahrs.pitch;
+
+        // rotate to earth-frame angles
```
Comment (djnugent) [inline libraries/AC_PrecLand/AC_PrecLand.cpp]: This need to be run despite the frame of ref(unless you have a 3 axis gimbal)

**Comment by rmackay9** [comment]
I've added the need-rework tag so we know the status until @squilter addresses @djnugent's comment

File: libraries/AC_PrecLand/AC_PrecLand_Companion.h [hunk 2]
```diff
@@ -37,7 +39,8 @@ class AC_PrecLand_Companion : public AC_PrecLand_Backend
 private:
 
     // output from camera
-    Vector2f            _bf_angle_to_target;    // last body-frame angle to target
+    MAV_FRAME           _frame;                 // what frame of reference is our sensor reporting in?
+    Vector2f            _angle_to_target;       // last body-frame angle to target
```
Comment (djnugent) [inline libraries/AC_PrecLand/AC_PrecLand_Companion.h]: Looks good. You may want to change the comments to reflect the change in variable names

**Comment by squilter** [comment]
@djnugent Does your simulator work?

**Comment by djnugent** [comment]
Sorry for the slow response time. My test setup was not up to date.

However, it does work. It is good to merge. @rmackay9 

Thanks @squilter

**Comment by squilter** [comment]
Thanks for the help, @djnugent

**Comment by squilter** [comment]
@rmackay9 ready, cap'n!

**Comment by rmackay9** [comment]
merged, thanks!

**Pull request closed without merging**